pkg/db/writer/store_test.go

```diff
@@ -174,17 +176,54 @@ func assertVulnerabilityMetadataReader(t *testing.T, reader db.VulnerabilityMeta
 	if actual, err := reader.GetVulnerabilityMetadata(id, recordSource); err != nil {
 		t.Fatalf("failed to get metadata: %+v", err)
 	} else {
-
-		diffs := deep.Equal(expected, actual)
-		if len(diffs) > 0 {
-			for _, d := range diffs {
-				t.Errorf("Diff: %+v", d)
+		sortMetadataCvss(actual.Cvss)
+		sortMetadataCvss(expected.Cvss)
+		for idx, actualCvss := range actual.Cvss {
+			assert.Equal(t, actualCvss.Vector, expected.Cvss[idx].Vector)
+			assert.Equal(t, actualCvss.Version, expected.Cvss[idx].Version)
+			assert.Equal(t, actualCvss.Metrics, expected.Cvss[idx].Metrics)
+
+			actualVendor, err := json.Marshal(actualCvss.VendorMetadata)
+			if err != nil {
+				t.Errorf("unable to marshal vendor metadata: %q", err)
 			}
+			expectedVendor, err := json.Marshal(expected.Cvss[idx].VendorMetadata)
+			if err != nil {
+				t.Errorf("unable to marshal vendor metadata: %q", err)
+			}
+			assert.Equal(t, string(actualVendor), string(expectedVendor))
+
 		}
+
+		assert.Equal(t, actual.ID, expected.ID)
+		assert.Equal(t, actual.RecordSource, expected.RecordSource)
+		assert.Equal(t, actual.Severity, expected.Severity)
+		assert.Equal(t, actual.Links, expected.Links)
+		assert.Equal(t, actual.Description, expected.Description)
```

Comment:
There is potential for misuse here (false positive tests) because I'm forced to test field by field since the VendorMetadata can be anything and the types aren't preserved when retrieving them back from the database. In the future, if more fields are added to the struct and not explicitly added here, then the test would still pass regardless of how correct/incorrect the values|fields are

Reply:
Also, I don't think I've quite wrapped my head around why testing field by field is necessary. It looks like `actual` and `expected` are of the same type. Which types aren't preserved?

Reply:
> Also, I don't think I've quite wrapped my head around why testing field by field is necessary. It looks like `actual` and `expected` are of the same type. Which types aren't preserved?

The `VulnerabilityMetadata` for both actual and expected are indeed the same. What is _not_ the same is the `VendorMetadata` field which maps to `interface{}`. When that is stored in the database as `type Foo struct` it _does not come back_ as `type Foo`. This situation forces the test to use known fields, and then treat `VendorMetadata` separately

Reply:
I think I see now... thanks! Would one approach be to do the comparison with the field you don't want to test factored out?

E.g.:

```golang
expected.Cvss = nil
actual.Cvss = nil
assert.Equal(t, expected, actual)
```

Reply:
@alfredodeza Just noting the important caveat from our Zoom call earlier:

We'd want to ensure that this `assert...` helper function didn't have any side effects, since that could impact the caller's test flow and test results in non-obvious ways. To that end, it's a problem that the `expected` parameter is a pointer type, because changes to values (e.g. the `... = nil` lines) will impact data beyond the scope of this function.

It seems easy enough to change the parameter to a _non-pointer_ type, which I would recommend anyway to avoid unnecessary complexity here.

Reply:
Hey @alfredodeza just checking in on this

Reply:
This will get updated, thanks for ensuring I saw it!

Reply:
This got updated